### PR TITLE
Implement basic pointer support

### PR DESCRIPTION
**Pull request by ajanthan-k**
- Completed groundwork for pointer management in context
- Support pointer declarations (+ initialisation) and assignments

- ___TODO___: Pointer arithmetic

### CODE REVIEW

**Comment by ajanthan-k** [comment]
@ivan-chong

File: src/ast/ast_assignment.cpp [hunk 1]
```diff
@@ -39,9 +39,24 @@ void Assignment::compile(std::ostream& os, Context& context, int destReg) const
     expr2->compile(os, context, right_reg);
 
     // at this point r-value is in right_reg
-    if (!expr1->isArray()) {
+    if(expr1->isPointer()) {
```
Comment (ivan-chong) [inline src/ast/ast_assignment.cpp]: Can we rename isPointer() and isPointer found on line 52, as they share the name but have different functionality

File: src/ast/ast_declaration.cpp [hunk 2]
```diff
@@ -30,7 +30,13 @@ void Declaration::compile(std::ostream& os, Context& context, int destReg) const
         // add to bindings with array name storing offset of a[0]
         context.addArray(var_name, type, array_size);
 
-    } else {
+    } else if (init_declarator->isPointer() == true) {
```
Comment (ivan-chong) [inline src/ast/ast_declaration.cpp]: Can we remove == true to be consistent with the rest of the codebase

File: src/ast/ast_functions.cpp [hunk 3]
```diff
@@ -165,14 +170,17 @@ void ParamDeclaration::compile(std::ostream& os, Context& context, int destReg)
...
 
-    int param_offset = context.addParam(var_name, type, param_index);
+    int param_offset = context.addParam(var_name, type, param_index, is_pointer);
 
     // params that didn't fit are stored at bottom of calller stack frame 0(sp), 4(sp)...
     // which is equivalent to callee 0(s0), 4(s0)
     if (param_offset == 1) return;
 
-    switch(type) {
+    Specifier effective_type = is_pointer ? Specifier::_int : type;
```
Comment (ivan-chong) [inline src/ast/ast_functions.cpp:183]: Can we comment to explain this and why we default to int

File: src/ast/ast_init_declarator.cpp [hunk 4]
```diff
@@ -47,6 +46,8 @@ void Init_Declarator::compile(std::ostream& os, Context& context, int destReg) c
         int init_reg;
         std::string store_instruction;
 
+        if(isPointer() == true) type = Specifier::_int;
```
Comment (ivan-chong) [inline src/ast/ast_init_declarator.cpp]: Similarly here == true is not necessary

**Review by ivan-chong — CHANGES_REQUESTED** [review]

**Review by ivan-chong — APPROVED** [review]

**Pull request merged**